File: app/src/main/java/com/example/dreamloaf/repository/SaleRepository.java
package com.example.dreamloaf.repository;

import android.app.Application;
import androidx.lifecycle.LiveData;
import com.example.dreamloaf.data.AppDatabase;
import com.example.dreamloaf.data.Product;
import com.example.dreamloaf.data.Sale;
import com.example.dreamloaf.data.SaleDao;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executors;

public class SaleRepository {
    private final SaleDao saleDao;
    private final AppDatabase database;

    public SaleRepository(Application application) {
        database = AppDatabase.getInstance(application);
        saleDao = database.saleDao();
    }

    public void addSale(int productId, int quantity, String date) {
        Executors.newSingleThreadExecutor().execute(() -> {
            Sale sale = new Sale();
            sale.setProductId(productId);
            sale.setQuantity(quantity);
            sale.setDate(date);
            sale.setUserId(1);
            saleDao.insert(sale);
        });
    }

    public void addSale(Sale sale) {
        Executors.newSingleThreadExecutor().execute(() -> saleDao.insert(sale));
    }

    public LiveData<List<Sale>> getSalesByDate(String date) {
        return saleDao.getSalesByDate(date);
    }

    public LiveData<List<Sale>> getAllSales() {
        return saleDao.getAllSales();
    }

    public LiveData<Map<String, Integer>> getMonthlySales() {
        return saleDao.getMonthlySales();
    }

    public LiveData<Double> getProductPrice(int productId) {
        return saleDao.getProductPrice(productId);
    }

    public void saveSale(int productId, int quantity, String date) {
        Executors.newSingleThreadExecutor().execute(() -> {
            Sale existing = saleDao.getSaleForProductAndDate(productId, date);
            if (existing != null) {
                existing.setQuantity(quantity);
                saleDao.update(existing);
            } else {
                Sale sale = new Sale();
                sale.setProductId(productId);
                sale.setQuantity(quantity);
                sale.setDate(date);
                sale.setUserId(1);
                saleDao.insert(sale);
            }
        });
    }

    public double calculateProfit(Sale sale, Product product) {
        if (product == null) return 0;
        return (product.getPrice() - product.getCostPrice()) * sale.getQuantity();
    }

    public double calculateTotalProfit(List<Sale> sales, List<Product> products) {
        if (sales == null || products == null) return 0;

        double totalProfit = 0;
        for (Sale sale : sales) {
            Product product = findProductById(products, sale.getProductId());
            if (product != null) {
                totalProfit += calculateProfit(sale, product);
            }
        }
        return totalProfit;
    }

    private Product findProductById(List<Product> products, int productId) {
        if (products == null) return null;
        for (Product product : products) {
            if (product.getId() == productId) {
                return product;
            }
        }
        return null;
    }
}

File: app/src/main/java/com/example/dreamloaf/repository/SaleRepository.kt
package com.example.dreamloaf.repository

import android.app.Application
import androidx.lifecycle.LiveData
import com.example.dreamloaf.data.AppDatabase
import com.example.dreamloaf.data.Product
import com.example.dreamloaf.data.Sale
import com.example.dreamloaf.data.SaleDao
import java.util.concurrent.Executors

class SaleRepository(application: Application?) {
    private val saleDao: SaleDao
    private val database: AppDatabase

    init {
        database = AppDatabase.getInstance(application!!.applicationContext)!!
        saleDao = database.saleDao()!!
    }

    fun addSale(productId: Int, quantity: Int, date: String?) {
        Executors.newSingleThreadExecutor().execute {
            val sale = Sale()
            sale.productId = productId
            sale.quantity = quantity
            sale.date = date
            sale.userId = 1
            saleDao.insert(sale)
        }
    }

    fun addSale(sale: Sale) {
        Executors.newSingleThreadExecutor().execute { saleDao.insert(sale) }
    }

    fun getSalesByDate(date: String?): LiveData<MutableList<Sale?>?>? {
        return saleDao.getSalesByDate(date)
    }

    val allSales: LiveData<MutableList<Sale?>?>?
        get() = saleDao.allSales

    val monthlySales: LiveData<MutableMap<String?, Int?>?>?
        get() = saleDao.getMonthlySales()

    fun getProductPrice(productId: Int): LiveData<Double?>? {
        return saleDao.getProductPrice(productId)
    }

    fun saveSale(productId: Int, quantity: Int, date: String?) {
        Executors.newSingleThreadExecutor().execute {
            val existing = saleDao.getSaleForProductAndDate(productId, date)
            if (existing != null) {
                existing.quantity = quantity
                saleDao.update(existing)
            } else {
                val sale = Sale()
                sale.productId = productId
                sale.quantity = quantity
                sale.date = date
                sale.userId = 1
                saleDao.insert(sale)
            }
        }
    }

    fun calculateProfit(sale: Sale, product: Product?): Double {
        if (product == null) return 0.0
        return (product.price - product.costPrice) * sale.quantity
    }

    fun calculateTotalProfit(sales: MutableList<Sale>?, products: MutableList<Product>?): Double {
        if (sales == null || products == null) return 0.0

        var totalProfit = 0.0
        for (sale in sales) {
            val product = findProductById(products, sale.productId)
            if (product != null) {
                totalProfit += calculateProfit(sale, product)
            }
        }
        return totalProfit
    }

    private fun findProductById(products: MutableList<Product>?, productId: Int): Product? {
        if (products == null) return null
        for (product in products) {
            if (product.id.toInt() == productId) {
                return product
            }
        }
        return null
    }
}